src/services/catalog/app.py

```diff
@@ -73,8 +74,30 @@ def GET(self, *path, **args):
 
         elif path[0] == "services":
 
+            if "type" in args:
+                # /catalog/services?type={service,device}&subtype={arduino,raspberry}
+                try:
+                    stype = ServiceType.value_of(str(args["type"]).upper())
+                except ValueError:
+                    cherrypy.response.status = 400
+                    return self.asjson_error("Invalid service type")
+
+                sers = [s for s in self._serviceManager.services.values() if s.stype == stype]
+                if "subtype" in args:
+                    try:
+                        subtype = ServiceSubType.value_of(str(args["subtype"]).upper())
+                    except ValueError:
+                        cherrypy.response.status = 400
+                        return self.asjson_error("Invalid service subtype")
+
+                    return self.asjson({"services": [s.toDict() for s in sers if s.subtype == subtype]})
+
+                # /catalog/services?type={service,device}
+                else:
+                    return self.asjson({"services": [s.toDict() for s in sers]})
+
```

Comment:
I don't get these lines of code. Could you please explain them? Thank you

Reply:
An entry in the catalog can be a Service or a Device. A device can be an Arduino or a Raspberry.

What I check if there is the optional argument type. If yes, I extract all the services of the specified type with  
```python
sers = [s for s in self._serviceManager.services.values() if s.stype == stype]
```

Then I check if there is the argument "subtype" specified in the url. If yes, I need to apply a new filter on top of the previous result with

```python
return self.asjson({"services": [s.toDict() for s in sers if s.subtype == subtype]})
```

Let me know if it's clear now!

Reply:
Let me know if this is ok too so we can resolve this conversation @giop98

Reply:
Sorry for the delay. Now it's more clear. Thank you @thegabriele97. I will resolve this conflict